File: server-csharp/Magician/Reducers.cs
using System.Diagnostics.Contracts;
using System.Numerics;
using System;
using SpacetimeDB;

public static partial class Module
{
    [Reducer]
    public static void HandleMovementRequestMagician(ReducerContext ctx, MovementRequest Request)
    {
        Magician Character = ctx.Db.magician.identity.Find(ctx.Sender) ?? throw new Exception("Magician To Move Not Found");

        Character.Rotation = Request.Aim;
        Character.Velocity = new DbVector3(0f, Character.Velocity.y, 0f);
        Character.KinematicInformation.Jump = false;

        if (GetPermissionEntry(Character.PlayerPermissionConfig, "CanWalk").Subscribers.Count == 0)
        {
            float LocalX = 0f;
            float LocalZ = 0f;

            if (Request.MoveForward && !Request.MoveBackward) LocalZ = 2f;
            else if (Request.MoveBackward && !Request.MoveForward) LocalZ = -2f;

            if (Request.MoveRight && !Request.MoveLeft) LocalX = 2f;
            else if (Request.MoveLeft && !Request.MoveRight) LocalX = -2f;

            if (GetPermissionEntry(Character.PlayerPermissionConfig, "CanRun").Subscribers.Count == 0 && Request.Sprint && Request.MoveForward && !Request.MoveBackward)
                LocalZ *= 2.5f;
            
            if (GetPermissionEntry(Character.PlayerPermissionConfig, "CanRun").Subscribers.Count == 0 && Request.Sprint)
                LocalX *= 1.5f;

            float YawRadians = ToRadians(Character.Rotation.Yaw);
            float CosYaw = MathF.Cos(YawRadians);
            float SinYaw = MathF.Sin(YawRadians);

            float WorldX = CosYaw * LocalX + SinYaw * LocalZ;
            float WorldZ = -SinYaw * LocalX + CosYaw * LocalZ;

            Character.Velocity = new DbVector3(WorldX, Character.Velocity.y, WorldZ);
        }

        if (GetPermissionEntry(Character.PlayerPermissionConfig, "CanJump").Subscribers.Count == 0 && Request.Jump)
        {
            Character.KinematicInformation.Jump = true;
            Character.Velocity.y = 7.5f;
        }

        if (GetPermissionEntry(Character.PlayerPermissionConfig, "CanCrouch").Subscribers.Count == 0 && Request.Crouch)
        {
            Character.Velocity = new DbVector3(Character.Velocity.x * 0.5f, Character.Velocity.y, Character.Velocity.z * 0.5f);
            Character.KinematicInformation.Crouched = true;
            AddSubscriberUnique(GetPermissionEntry(Character.PlayerPermissionConfig, "CanRun").Subscribers, "Crouch");
        }

        if (Request.Crouch is false)
        {
            Character.KinematicInformation.Crouched = false;
            RemoveSubscriber(GetPermissionEntry(Character.PlayerPermissionConfig, "CanRun").Subscribers, "Crouch");
        }

        ctx.Db.magician.identity.Update(Character);
    }   

    [Reducer]
    public static void HandleActionChangeRequestMagician(ReducerContext Ctx, ActionRequestMagician Request)
    {
        Identity Identity = Ctx.Sender;
        Magician Character = Ctx.Db.magician.identity.Find(Identity) ?? throw new Exception("Magician Not Found");
        MagicianState OldState = Character.State;

        if (Request.State is MagicianState.Attack && GetPermissionEntry(Character.PlayerPermissionConfig, "CanAttack").Subscribers.Count == 0 && Character.Bullets.Count > 0)
        {
            Character.State = MagicianState.Attack;
            AddSubscriberUnique(GetPermissionEntry(Character.PlayerPermissionConfig, "CanAttack").Subscribers, "Attack");
            AddSubscriberUnique(GetPermissionEntry(Character.PlayerPermissionConfig, "CanReload").Subscribers, "Attack");
            TryPerformAttack(Ctx, ref Character, Request.AttackInformation);
        }

        else if (Request.State is MagicianState.Reload && GetPermissionEntry(Character.PlayerPermissionConfig, "CanReload").Subscribers.Count == 0 && Character.Bullets.Count < Character.BulletCapacity)
        {
            Character.State = MagicianState.Reload;
            AddSubscriberUnique(GetPermissionEntry(Character.PlayerPermissionConfig, "CanReload").Subscribers, "Reload");
        }

        bool StateSwitched = OldState != Character.State;
        if (StateSwitched)
        {
            ResetAllTimers(ref Character);
            switch (OldState)
            {
                case MagicianState.Attack:
                    RemoveSubscriber(GetPermissionEntry(Character.PlayerPermissionConfig, "CanAttack").Subscribers, "Attack");
                    RemoveSubscriber(GetPermissionEntry(Character.PlayerPermissionConfig, "CanReload").Subscribers, "Attack");
                    break;

                case MagicianState.Reload:
                    RemoveSubscriber(GetPermissionEntry(Character.PlayerPermissionConfig, "CanReload").Subscribers, "Reload");
                    break;

                default:
                    break;
            }
        }

        Ctx.Db.magician.identity.Update(Character);
    }

    [Reducer]
    public static void HandleMagicianTimers(ReducerContext Ctx, Handle_Magician_Timers_Timer timer)
    {
        float Time = timer.tick_rate;
        foreach (Magician magician in Ctx.Db.magician.MatchId.Filter(timer.MatchId))
        {
            // If Certain Timers Ever Get Introduced That Are Irrelevant To Magician State, Seperate Timers Into Two Properties: Stateless and Stateful Timers, One That Relies On The State To Know Which Timer To Adjust, One That Adjusts All The Timers Regardless Of State. This Would Be The Stateful Timer
            Magician Magician = magician;
            Timer Timer;
            switch (Magician.State)
            {
                case MagicianState.Attack:
                    Timer = Magician.Timers[TryFindTimerIndex(ref Magician, "Attack")];
                    Timer.CurrentTime -= Time;

                    if (Timer.CurrentTime <= 0) 
                    {
                        Timer.CurrentTime = Timer.ResetTime;
                        if (Magician.Bullets.Count > 0)
                            Magician.State = MagicianState.Default;
                        
                        else {
                            Magician.State = MagicianState.Reload;
                            AddSubscriberUnique(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanReload").Subscribers, "Reload");
                        }

                        RemoveSubscriber(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanAttack").Subscribers, "Attack");
                        RemoveSubscriber(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanReload").Subscribers, "Attack");  
                    }

                    Magician.Timers[TryFindTimerIndex(ref Magician, "Attack")] = Timer;
                    break;
                
                case MagicianState.Reload:
                    Timer = Magician.Timers[TryFindTimerIndex(ref Magician, "Reload")];
                    Timer.CurrentTime -= Time;

                    if (Timer.CurrentTime <= 0) 
                    {
                        Timer.CurrentTime = Timer.ResetTime;
                        Magician.State = MagicianState.Default;
                        RemoveSubscriber(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanReload").Subscribers, "Reload");
                        TryReload(Ctx, ref Magician);
                    }

                    Magician.Timers[TryFindTimerIndex(ref Magician, "Reload")] = Timer;
                    break;

                case MagicianState.Default:
                    break;

                default:
                    break;
            }

            Ctx.Db.magician.identity.Update(Magician);
        }
    }

    [Reducer]
    public static void ApplyGravityMagician(ReducerContext ctx, Gravity_Timer_Magician Timer)
    {
        var Time = Timer.tick_rate;
        foreach (var charac in ctx.Db.magician.MatchId.Filter(Timer.MatchId))
        {
            var character = charac;
            character.Velocity.y = character.Velocity.y > -10f ? character.Velocity.y -= Timer.gravity * Time : -10f;
            ctx.Db.magician.identity.Update(character);
        }
    }
    
    [Reducer]
    public static void AddCollisionEntryMagician(ReducerContext ctx, CollisionEntry Entry, Identity TargetIdentity)
    {
        Magician Magician = ctx.Db.magician.identity.Find(TargetIdentity) ?? throw new Exception("Magician (Sender) Not Found");
        if (Magician.CollisionEntries.Contains(Entry) is false) Magician.CollisionEntries.Add(Entry);
        ctx.Db.magician.identity.Update(Magician);
    }

    [Reducer]
    public static void RemoveCollisionEntryMagician(ReducerContext ctx, CollisionEntry Entry, Identity TargetIdentity)
    {
        Magician Magician = ctx.Db.magician.identity.Find(TargetIdentity) ?? throw new Exception("Magician (Sender) Not Found");
        if (Magician.CollisionEntries.Contains(Entry) is true) Magician.CollisionEntries.Remove(Entry);
        ctx.Db.magician.identity.Update(Magician);
    }

    [Reducer]
    public static void MoveMagicians(ReducerContext Ctx, Move_All_Magicians_Timer Timer)
    {
        float TickTime = Timer.tick_rate;
        float MinTimeStep = 1e-4f;
        int MaxSubsteps = 4;

        foreach (var CharacterRow in Ctx.Db.magician.MatchId.Filter(Timer.MatchId))
        {
            Magician Character = CharacterRow;

            bool WasGrounded = Character.KinematicInformation.Grounded;
            Character.KinematicInformation.Grounded = false;
            Character.IsColliding = false;
            Character.CorrectedVelocity = Character.Velocity;

            List<CollisionContact> PreContacts = new List<CollisionContact>();
            foreach (CollisionEntry Entry in Character.CollisionEntries)
            {
                TryBuildContactForEntry(Ctx, ref Character, Entry, PreContacts);
            }

            if (PreContacts.Count > 0)
            {
                ResolveContacts(ref Character, PreContacts, Character.Velocity);
            }

            float RemainingTime = TickTime;
            int SubstepCount = 0;

            while (RemainingTime > MinTimeStep && SubstepCount < MaxSubsteps)
            {
                SubstepCount += 1;
                float StepTime = RemainingTime / (MaxSubsteps - SubstepCount + 1);
                DbVector3 StepVelocity = Character.IsColliding ? Character.CorrectedVelocity : Character.Velocity;

                Character.Position = new DbVector3(
                    Character.Position.x + StepVelocity.x * StepTime,
                    Character.Position.y + StepVelocity.y * StepTime,
                    Character.Position.z + StepVelocity.z * StepTime
                );

                foreach (CollisionEntry Entry in Character.CollisionEntries)
                {
                    if (TryForceOverlapForEntry(Ctx, ref Character, Entry, WasGrounded) is true)
                        break;
                }

                List<CollisionContact> PostContacts = new List<CollisionContact>();
                foreach (CollisionEntry Entry in Character.CollisionEntries)
                {
                    TryBuildContactForEntry(Ctx, ref Character, Entry, PostContacts);
                }

                if (PostContacts.Count > 0)
                {
                    ResolveContacts(ref Character, PostContacts, Character.Velocity);
                }

                RemainingTime -= StepTime;
            }

            DbVector3 FinalStepVelocity = Character.IsColliding ? Character.CorrectedVelocity : Character.Velocity;

            float GroundStickVelocityThreshold = 2f;
            bool GroundedThisTick = Character.KinematicInformation.Grounded;

            if (GroundedThisTick is false && WasGrounded is true && MathF.Abs(FinalStepVelocity.y) < GroundStickVelocityThreshold)
                Character.KinematicInformation.Grounded = true;

            AdjustGrounded(Ctx, FinalStepVelocity, ref Character);

            Ctx.Db.magician.identity.Update(Character);
        }
    }
    
}

File: server-rust/spacetimedb/src/Magicians/Reducers.rs
use spacetimedb::{reducer, Identity, ReducerContext, Table};
use crate::*;

#[reducer]
pub fn handle_movement_request_magician(ctx: &ReducerContext, request: MovementRequest) 
{
    let mut character = ctx.db.magician().identity().find(ctx.sender).expect("Magician To Move Not Found");

    character.rotation = request.aim;
    character.velocity = DbVector3 { x: 0.0, y: character.velocity.y, z: 0.0 };
    character.kinematic_information.jump = false;
    let speed_mutiplier = character.combat_information.speed_multiplier;
    let stunned = is_permission_unblocked(&character.permissions, "Stunned") == false;
    let taroted = is_permission_unblocked(&character.permissions, "Taroted") == false;

    if is_permission_unblocked(&character.permissions, "CanWalk") && stunned == false {
        let mut local_x: f32 = 0.0;
        let mut local_z: f32 = 0.0;

        if request.move_forward && !request.move_backward {
            local_z = 2.0;
        } 
        
        else if request.move_backward && !request.move_forward {
            local_z = -2.0;
        }

        if request.move_right && !request.move_left {
            local_x = 2.0;
        } 
        
        else if request.move_left && !request.move_right {
            local_x = -2.0;
        }

        if is_permission_unblocked(&character.permissions, "CanRun") && request.sprint && request.move_forward && !request.move_backward{
            local_z *= 2.5;
        }

        if is_permission_unblocked(&character.permissions, "CanRun") && request.sprint {
            local_x *= 1.5;
        }

        let yaw_radians: f32 = to_radians(character.rotation.yaw);
        let cos_yaw: f32 = yaw_radians.cos();
        let sin_yaw: f32 = yaw_radians.sin();

        let world_x: f32 = cos_yaw * local_x + sin_yaw * local_z;
        let world_z: f32 = -sin_yaw * local_x + cos_yaw * local_z;

        character.velocity = DbVector3 { x: world_x, y: character.velocity.y, z: world_z };
    }

    if is_permission_unblocked(&character.permissions, "CanJump") && request.jump && stunned == false {
        character.kinematic_information.jump = true;
        character.velocity.y = 7.5;
    }

    if is_permission_unblocked(&character.permissions, "CanCrouch") && request.crouch && stunned == false {
        character.velocity = DbVector3 { x: character.velocity.x * 0.5, y: character.velocity.y, z: character.velocity.z * 0.5 };
        character.kinematic_information.crouched = true;
        add_subscriber_to_permission(&mut character.permissions, "CanRun", "Crouch");
    }

    if !request.crouch || stunned == true {
        character.kinematic_information.crouched = false;
        remove_subscriber_from_permission(&mut character.permissions, "CanRun", "Crouch");
    }

    character.velocity = if taroted { DbVector3 { x: character.velocity.x * speed_mutiplier * -1.0, y: character.velocity.y, z: character.velocity.z * speed_mutiplier -1.0 } } else { DbVector3 { x: character.velocity.x * speed_mutiplier, y: character.velocity.y, z: character.velocity.z * speed_mutiplier } };
    ctx.db.magician().identity().update(character);
}

#[reducer]
pub fn handle_action_change_request_magician(ctx: &ReducerContext, request: ActionRequestMagician) 
{
    let mut magician = ctx.db.magician().identity().find(ctx.sender).expect("Magician Not Found");
    let stunned = is_permission_unblocked(&magician.permissions, "Stunned") == false;
    let old_state: MagicianState = magician.state;

    if request.state == MagicianState::Attack && is_permission_unblocked(&magician.permissions, "CanAttack") && magician.bullets.len() > 0 && stunned == false {
        magician.state = MagicianState::Attack;
        add_subscriber_to_permission(&mut magician.permissions, "CanAttack", "Attack");
        add_subscriber_to_permission(&mut magician.permissions, "CanReload", "Attack");
        add_subscriber_to_permission(&mut magician.permissions, "CanDust", "Attack");
        add_subscriber_to_permission(&mut magician.permissions, "CanCloak", "Attack");
        add_subscriber_to_permission(&mut magician.permissions, "CanHypnosis", "Attack");
        try_perform_attack(ctx, &mut magician, request.attack_information);
    } 
    
    else if request.state == MagicianState::Reload && is_permission_unblocked(&magician.permissions, "CanReload") && (magician.bullets.len() as i32) < magician.bullet_capacity && stunned == false {
        magician.state = MagicianState::Reload;
        add_subscriber_to_permission(&mut magician.permissions, "CanReload", "Reload");
    }

    else if request.state == MagicianState::Dust && is_permission_unblocked(&magician.permissions, "CanDust") && stunned == false {
        magician.state = MagicianState::Dust;
        add_subscriber_to_permission(&mut magician.permissions, "CanDust", "Dust");
        add_subscriber_to_permission(&mut magician.permissions, "CanAttack", "Dust");
        add_subscriber_to_permission(&mut magician.permissions, "CanReload", "Dust");
        add_subscriber_to_permission(&mut magician.permissions, "CanCloak", "Dust");
        add_subscriber_to_permission(&mut magician.permissions, "CanHypnosis", "Dust");
        try_perform_dust(ctx, &mut magician, request.dust_information)
    }
    
    else if request.state == MagicianState::Cloak && is_permission_unblocked(&magician.permissions, "CanCloak") && stunned == false {
        magician.state = MagicianState::Cloak;
        add_subscriber_to_permission(&mut magician.permissions, "CanCloak", "Cloak");
    }

    else if request.state == MagicianState::Hypnosis && is_permission_unblocked(&magician.permissions, "CanHypnosis") && stunned == false {
        magician.state = MagicianState::Hypnosis;
        add_subscriber_to_permission(&mut magician.permissions, "CanHypnosis", "Hypnosis");
        add_subscriber_to_permission(&mut magician.permissions, "CanDust", "Hypnosis");
        add_subscriber_to_permission(&mut magician.permissions, "CanAttack", "Hypnosis");
        add_subscriber_to_permission(&mut magician.permissions, "CanReload", "Hypnosis");
        add_subscriber_to_permission(&mut magician.permissions, "CanCloak", "Hypnosis");
    }

    if old_state != magician.state {
        adjust_timer_for_interruptable_state(&mut magician, old_state);
        match old_state {
            MagicianState::Reload => {
                remove_subscriber_from_permission(&mut magician.permissions, "CanReload", "Reload");
            }

            MagicianState::Cloak => { }

            _ => {}
        }
    }

    if magician.state != MagicianState::Default && magician.state != MagicianState::Reload && magician.state != MagicianState::Cloak {
        try_interrupt_cloak_and_speed_effects_magician(ctx, &mut magician);
    }

    ctx.db.magician().identity().update(magician);
}

#[reducer]
pub fn handle_stateless_action_request_magician(ctx: &ReducerContext, request: StatelessActionRequestMagician) 
{
    let mut magician = ctx.db.magician().identity().find(ctx.sender).expect("Magician Not Found");

    if request.action == MagicianStatelessAction::Tarot && is_permission_unblocked(&magician.permissions, "CanTarot") {
        try_tarot(ctx, &mut magician);
        add_subscriber_to_permission(&mut magician.permissions, "CanTarot", "Tarot");
    }

    ctx.db.magician().identity().update(magician);
}

#[reducer]
pub fn handle_magician_timers(ctx: &ReducerContext, timer: HandleMagicianTimersTimer) {
    let time: f32 = timer.tick_rate;
    for mut magician in ctx.db.magician().game_id().filter(timer.game_id) {
        match magician.state {
            MagicianState::Attack => {
                if tick_active_timer_and_check_expired(&mut magician, "Attack", time) {
                    if magician.bullets.len() > 0 {
                        magician.state = MagicianState::Default;
                    } 
                    
                    else {
                        magician.state = MagicianState::Reload;
                        add_subscriber_to_permission(&mut magician.permissions, "CanReload", "Reload"); // Add To Every Other State To Ensure Transition Back To Reload If Necessary
                    }

                    remove_subscriber_from_permission(&mut magician.permissions, "CanReload", "Attack");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanDust", "Attack");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanCloak", "Attack");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanHypnosis", "Attack");
                }
            }

            MagicianState::Reload => {  
                if tick_active_timer_and_check_expired(&mut magician, "Reload", time) {
                    magician.state = MagicianState::Default;
                    try_reload(ctx, &mut magician);
                }
            }

            MagicianState::Dust => {
                if tick_active_timer_and_check_expired(&mut magician, "Dust", time) {
                    magician.state = MagicianState::Default;
                    remove_subscriber_from_permission(&mut magician.permissions, "CanReload", "Dust");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanAttack", "Dust");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanCloak", "Dust");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanHypnosis", "Dust");
                }
            }

            MagicianState::Cloak => {
                if tick_active_timer_and_check_expired(&mut magician, "Cloak", time) {
                    magician.state = MagicianState::Default;
                    try_cloak(ctx, &mut magician);
                }
            }

            MagicianState::Hypnosis => {
                if tick_active_timer_and_check_expired(&mut magician, "Hypnosis", time) {
                    magician.state = MagicianState::Default;
                    remove_subscriber_from_permission(&mut magician.permissions, "CanReload", "Hypnosis");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanAttack", "Hypnosis");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanCloak", "Hypnosis");
                    remove_subscriber_from_permission(&mut magician.permissions, "CanDust", "Hypnosis");
                    try_hypnosis(ctx, &mut magician);
                }
            }

            MagicianState::Default => {}
        }

        for i in 0..magician.timers.len() {
            if let Some(expired_timer_name) = tick_cooldown_timer_and_check_expired(&mut magician.timers[i], time) {
                match expired_timer_name.as_str() {
                    "Attack" => remove_subscriber_from_permission(&mut magician.permissions, "CanAttack", "Attack"),
                    "Reload" => remove_subscriber_from_permission(&mut magician.permissions, "CanReload", "Reload"),
                    "Dust" => remove_subscriber_from_permission(&mut magician.permissions, "CanDust", "Dust"),
                    "Cloak" => remove_subscriber_from_permission(&mut magician.permissions, "CanCloak", "Cloak"),
                    "Hypnosis" => remove_subscriber_from_permission(&mut magician.permissions, "CanHypnosis", "Hypnosis"),
                    _ => {}
                }
            }
        }

        ctx.db.magician().identity().update(magician);
    }
}

#[reducer]
pub fn handle_magician_stateless_timers(ctx: &ReducerContext, timer: HandleMagicianStatelessTimersTimer) 
{ 
    let time: f32 = timer.tick_rate;
    for mut magician in ctx.db.magician().game_id().filter(timer.game_id) { 
        for i in 0..magician.stateless_timers.len() {
            if let Some(expired_timer_name) = tick_stateless_cooldown_timer_and_check_expired(&mut magician.stateless_timers[i], time) { 
                match expired_timer_name.as_str() {
                    "Tarot" => remove_subscriber_from_permission(&mut magician.permissions, "CanTarot", "Tarot"),
                    _ => {}

                }
            }
        }
        
        ctx.db.magician().identity().update(magician);
    }
}

#[reducer]
pub fn apply_gravity_magician(ctx: &ReducerContext, timer: GravityTimerMagician) 
{
    let time: f32 = timer.tick_rate;
    for character_row in ctx.db.magician().game_id().filter(timer.game_id) {
        let mut character = character_row;

        if character.velocity.y > -10.0 { 
            character.velocity.y -= timer.gravity * time; 
        }

        else { 
            character.velocity.y = -10.0; 
        }

        ctx.db.magician().identity().update(character);
    }
}

#[reducer]
pub fn add_collision_entry_magician(ctx: &ReducerContext, entry: CollisionEntry, target_identity: Identity) 
{
    let mut magician = ctx.db.magician().identity().find(target_identity).expect("Magician (Sender) Not Found");
    if magician.collision_entries.contains(&entry) == false { 
        magician.collision_entries.push(entry); 
        ctx.db.magician().identity().update(magician);
    }  
}

#[reducer]
pub fn remove_collision_entry_magician(ctx: &ReducerContext, entry: CollisionEntry, target_identity: Identity) 
{
    let mut magician = ctx.db.magician().identity().find(target_identity).expect("Magician (Sender) Not Found");
    if let Some(index) = magician.collision_entries.iter().position(|existing| *existing == entry) {
        magician.collision_entries.swap_remove(index);
        ctx.db.magician().identity().update(magician);
    }
}

#[reducer]
pub fn move_magicians(ctx: &ReducerContext, timer: MoveAllMagiciansTimer) 
{
    let tick_time: f32 = timer.tick_rate;
    let min_time_step: f32 = 1e-4;
    let max_substeps: i32 = 4;

    for mut magician in ctx.db.magician().game_id().filter(timer.game_id) {
        let was_grounded: bool = magician.kinematic_information.grounded;
        magician.kinematic_information.grounded = false;
        magician.is_colliding = false;
        magician.corrected_velocity = magician.velocity;

        let mut pre_contacts: Vec<CollisionContact> = Vec::new();
        for entry in magician.collision_entries.iter() {
            try_build_contact_for_entry(ctx, &magician, entry, &mut pre_contacts);
        }

        if pre_contacts.is_empty() == false {
            let input_velocity = magician.velocity;
            resolve_contacts(&mut magician, &pre_contacts, input_velocity);
        }

        let mut remaining_time: f32 = tick_time;
        let mut substep_count: i32 = 0;

        let mut post_contacts: Vec<CollisionContact> = Vec::new();

        while remaining_time > min_time_step && substep_count < max_substeps {
            substep_count += 1;
            let step_time: f32 = remaining_time / ((max_substeps - substep_count + 1) as f32);
            let step_velocity = if magician.is_colliding { magician.corrected_velocity } else { magician.velocity };

            magician.position = add(magician.position, mul(step_velocity, step_time));

            let collision_entry_count: usize = magician.collision_entries.len();
            for entry_index in 0..collision_entry_count {
                let entry: CollisionEntry = magician.collision_entries[entry_index];
                if try_force_overlap_for_entry(ctx, &mut magician, &entry, was_grounded) {
                    break;
                }
            }

            post_contacts.clear();
            for entry in magician.collision_entries.iter() {
                try_build_contact_for_entry(ctx, &magician, entry, &mut post_contacts);
            }

            if post_contacts.is_empty() == false {
                let input_velocity = magician.velocity;
                resolve_contacts(&mut magician, &post_contacts, input_velocity);
            }

            remaining_time -= step_time;
        }

        let final_step_velocity = if magician.is_colliding { magician.corrected_velocity } else { magician.velocity };

        let ground_stick_velocity_threshold: f32 = 2.0;
        let grounded_this_tick: bool = magician.kinematic_information.grounded;

        if grounded_this_tick == false && was_grounded && final_step_velocity.y.abs() < ground_stick_velocity_threshold {
            magician.kinematic_information.grounded = true;
        }

        adjust_grounded(ctx, was_grounded, &final_step_velocity, &mut magician);
        ctx.db.magician().identity().update(magician);
    }
}

#[reducer]
pub fn move_magicians_lag_test(ctx: &ReducerContext, timer: MoveAllMagiciansTimer)
{
    let delta_time: f32 = timer.tick_rate;

    for mut magician in ctx.db.magician().game_id().filter(timer.game_id) {
        magician.position.x += magician.velocity.x * delta_time;
        magician.position.y += magician.velocity.y * delta_time;
        magician.position.z += magician.velocity.z * delta_time;

        if magician.position.y < 0.0 {
            magician.position.y = 0.0;

            if magician.velocity.y < 0.0 {
                magician.velocity.y = 0.0;
            }
        }

        ctx.db.magician().identity().update(magician);
    }
}

#[reducer]
pub fn hypnotise(ctx: &ReducerContext, camera_info: HypnosisCameraInformation)
{
    let mut magician = ctx.db.magician().identity().find(ctx.sender).expect("Magician Not Found!");

    let mut hypnosis_iterator = ctx.db.player_effects().target_and_type().filter((magician.id, EffectType::Hypnosis));
    let mut hypnosis_effect = match (hypnosis_iterator.next(), hypnosis_iterator.next()) {
        (None, _) => { return; },
        (Some(effect), None) => effect,
        (Some(_), Some(_)) => panic!("Target Magician Should Only Have One Hypnosis Effect At Most!"),
    };

    let raycast = try_hypnotise(ctx, &mut magician, camera_info);
    let raycast_target_id_option: Option<u64> = match raycast.hit_type {
        RaycastHitType::Magician => Some(raycast.hit_entity_id),
        _ => None,
    };

    let hypnosis_information = hypnosis_effect.hypnosis_informaton.as_mut().expect("Hypnosis Effect Must Have Hypnosis Information");
    let last_target_id_option: Option<u64> = hypnosis_information.last_target_id;

    match (last_target_id_option, raycast_target_id_option) {
        (Some(last_target_id), Some(raycast_target_id)) if last_target_id == raycast_target_id => { } // Last Target Id And Raycast Target Id Are Both Not None And Are The Same Id Value

        (Some(last_target_id), Some(raycast_target_id)) => { // Last Target Id And Raycast Target Id Are Both Not None But Are Different Id Values
            let mut stunned_magician = ctx.db.magician().id().find(last_target_id).expect("Stunned Magician Must Exist!");
            let mut stunned_iterator = ctx.db.player_effects().target_sender_and_type().filter((last_target_id, magician.id, EffectType::Stunned));

            let stunned_effect = match (stunned_iterator.next(), stunned_iterator.next()) {
                (None, _) => panic!("Stunned Iterator Should Have First Element!"),
                (Some(effect), None) => effect,
                (Some(_), Some(_)) => panic!("Target Magician Should Only Have One Stun Effect From Sender At Most!"),
            };

            undo_and_delete_stunned_effect_magician(ctx, &mut stunned_magician, stunned_effect.id);
            ctx.db.magician().id().update(stunned_magician);

            add_effects_to_table(ctx, vec![create_stunned_effect()], raycast_target_id, magician.id, magician.game_id);
            hypnosis_information.last_target_id = Some(raycast_target_id);
        }

        (None, Some(raycast_target_id)) => { // Last Target Id Is None But Raycast Target Has Id Value
            add_effects_to_table(ctx, vec![create_stunned_effect()], raycast_target_id, magician.id, magician.game_id);
            hypnosis_information.last_target_id = Some(raycast_target_id);
        }

        (Some(last_target_id), None) => { // Last Target Id Has Id Value But Raycast Target Is None
            let mut stunned_magician = ctx.db.magician().id().find(last_target_id).expect("Stunned Magician Must Exist!");
            let mut stunned_iterator = ctx.db.player_effects().target_sender_and_type().filter((last_target_id, magician.id, EffectType::Stunned));

            let stunned_effect = match (stunned_iterator.next(), stunned_iterator.next()) {
                (None, _) => panic!("Stunned Iterator Should Have First Element!"),
                (Some(effect), None) => effect,
                (Some(_), Some(_)) => panic!("Target Magician Should Only Have One Stun Effect From Sender At Most!"),
            };

            undo_and_delete_stunned_effect_magician(ctx, &mut stunned_magician, stunned_effect.id);
            ctx.db.magician().id().update(stunned_magician);

            hypnosis_information.last_target_id = None;
        }

        (None, None) => { } // Both Are None
    }

    ctx.db.player_effects().id().update(hypnosis_effect);
}


